Remove newlines by default when exporting contents using `secrets env`

aryanjassal opened the pull request:
### Description
<!-- Write your description about what this PR is about. -->

The standard to end all files in Unix is adding a newline at the end of the file. All editors, including `vim` do it. However, adding a stray newline at the end of a secret, like a password, can have adverse effects.

As such, this PR will, by default, trim the single trailing newline, with an option to disable it if needed.

### Issues Fixed
<!-- List all issues fixed by this PR. -->
* Fixes #296 (REF ENG-416)
* Related to #325 (REF ENG-466)

### Tasks
<!-- 
  List all tasks to be done by this PR.
  If a task is no longer required, add a strikethrough (including the checkbox):
  - ~~[ ] 3. ...~~ - being completed in #...
-->
- [x] 1. Remove newlines by default
- [x] 2. Add a new option to allow preserving newlines
- [x] 3. Add tests ensuring this behaviour

### Final checklist
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

aryanjassal commented:
We are using `spawnSync` to spawn the env command. This will result in everything else being blocked, and in cases of a long-running command, will also lead to websocket timeouts. To avoid this, I will change it to use `spawn` much like what I did in `secrets edit`.

Also, there was a minor inaccuracy in the command, where the command would fail without any environment variables being provided. This is different than the Unix `env`, which just runs the command with all the system variables instead. was this intentional or should I fix it?

@tegefaulkes @CMCDragonkai

CMCDragonkai commented:
In our case `polykey secrets env` doesn't make sense to run with no variables. But... What do you mean by runs with system variables?

aryanjassal commented:
> In our case `polykey secrets env` doesn't make sense to run with no variables. But... What do you mean by runs with system variables?

By default, when no variables are provided to `env`, it exports all the current shell variables into the command. This can be previewed by running `env` in the terminal, which would output all the current variables. That's what gets passed on to any command. Any other variables we define gets appended to that list.

In our case, it does not make much sense, but is a deviation from the standard Unix behaviour, which is why I wanted to ask and confirm this.

aryanjassal commented:
> We are using `spawnSync` to spawn the env command. This will result in everything else being blocked, and in cases of a long-running command, will also lead to websocket timeouts. To avoid this, I will change it to use `spawn` much like what I did in `secrets edit`.

After talking about this with Brian, I have realised that the Polykey client is actually stopped before the commands are even run. This means that, without a connection to Polykey, the websocket errors won't happen, and that is a non-issue in this case.

I did a manual sanity check by using a long-running command in `secrets env`, and it didn't throw any errors.

aryanjassal commented:
The syntax to preserve newlines for a secret is as follows:

```console
[aryanj@matrix-34xx:~]$ polykey secrets env vault:SECRET1 vault:SECRET2
SECRET1='testing123'
SECRET2='testing123'

[aryanj@matrix-34xx:~]$ polykey secrets env -pn vault:SECRET1 vault:SECRET2
SECRET1='testing123
'
SECRET2='testing123'

[aryanj@matrix-34xx:~]$ polykey secrets env --preserve-newline VAULT:SECRET3
SECRET3='newlines-after-this

'

[aryanj@matrix-34xx:~]$ polykey secrets env VAULT:SECRET3
SECRET3='newlines-after-this
'
```

If the newlines should be preserved, then the flag `-pn --preserve-newline` should be used before providing a secret.

aryanjassal commented:
How should I handle providing the same secret in two different contexts? If `SECRET1` has been provided with both a preserve newline and without, what should take precedence?

What about cases where we export entire vaults? Currently, the flag does preserve newlines of all the secrets exported from the vault, but what if we need one secret to have newlines, and the rest can be trimmed? Or vice-versa, where all secrets need to have preserved newlines, but one secret needs to be trimmed? How would be handle those cases?

With the way how the command is currently structured, is it not possible to preserve the order of the secrets, so they can't be used for prioritisation.

Currently, the preserving newline takes precedence over trimming them. For example, the following will preserve newlines for all secrets:

```console
[aryanj@matrix-34xx:~]$ polykey secrets env -pn vault vault:SECRET1
SECRET1='testing123
'
SECRET2='testing123
'
```

Should this be the intentional behaviour? @CMCDragonkai

aryanjassal commented on src/secrets/CommandEnv.ts:
```diff
@@ -33,7 +35,23 @@ class CommandEnv extends CommandPolykey {
         options,
       ) => {
-        args[1].shift();
+        // The parser sets the default value for args. If no arguments are
+        // provided, then args is an empty array []. This is different from the
+        // expected structure. This block ensure that the structure is preserved.
+        // Do the same for options.preserveNewline.
+        args = args[0] == null && args[1] == null ? [[], []] : args;
+        options.preserveNewline = options.preserveNewline ?? [];
+
+        // There needs to be at least one argument or preserved argument
+        // provided. Otherwise, print the help text and error out.
+        if (args[0].length === 0 && options.preserveNewline.length === 0) {
+          this.outputHelp()
+          throw new commander.InvalidArgumentError(
+            'You must provide at least one secret path.',
+          );
+        }
```
I don't like doing it this way, but I don't know any other way to make this work. I have set the newline preservation as an option. This means that we have zero secrets provided only when both `args` and `options.preserveNewline` are zero.

If there were no args, then the array will not get properly initialised. The same is the case for `options.preserveNewline`, so I had to manually create an empty array for both.

Then, `this.help()` will print the help page and exit with an error code of zero, so that can't be used for printing help text for an error, and the help text isn't printed for errors thrown from the action. It is only printed when it is thrown in the parsing step. So, I have to manually output the help page, and exit with the error so the correct error code gets set.

This approach just doesn't sit well with me, and there is got to be a better method to do this.

aryanjassal replied:
Answered in https://github.com/MatrixAI/Polykey-CLI/pull/329#issuecomment-2472332972

aryanjassal commented on src/utils/parsers.ts:
```diff
@@ -208,6 +208,8 @@ function parseEnvArgs(
+  console.error('cur', current)
   if (current[1].length === 0) {
     // Parse a secret path
     if (value !== '--') {
```
I have tried to think up of many solutions, but I can't find a way to consistently use `'--'` as a delimiter between the secrets and a command using commander. The simplest way to fix this would be to use commander to process and validate the arguments, and use `process.env` to get the raw arguments, which will include the `'--'`. We can parse the raw arguments and use them reliably, but I don't like the fact that we are basically bypassing the literal purpose of commander just to get something like this working.

This is kind of broken right now due to the option for removing newlines. If there are zero arguments, but the command doesn't fail, that means that there were some paths provided with the option to preserve the newlines. However, in this case, the `this.passThroughOptions()` doesn't pass through `'--'`, as it only does that after we have started listing the arguments, which didn't happen in this case, as we are listing individual paths inside the preserve newline option, so commander gobbles up the `'--'`. This can be seen in this command:

```console
[aryanj@matrix-34xx:~]$ polykey secrets env --preserve-newline vault:SECRET -- command
# args = [['command', undefined, undefined], []]
# the -- gets parsed by commander instead of the parser
# now the program thinks that 'command' is a vault name
```

We need to address this problem in this PR, but I'm not sure how I can even approach this. I need inputs.

aryanjassal replied:
Related issue: https://github.com/MatrixAI/Polykey-CLI/issues/325

aryanjassal replied:
Answered in https://github.com/MatrixAI/Polykey-CLI/pull/329#issuecomment-2472332972

aryanjassal reviewed (commented):
I've run into a roadblock and need some inputs from @CMCDragonkai and @tegefaulkes to help me with some issues.

CMCDragonkai commented:
You can have a global flag or the flags are applied per spec. I would do something like that.

CMCDragonkai commented:
By default secret env is going to need to trim a single trailing newline.

Repeating the spec is an interesting situation I would argue that you maintain order in line with whatever the order of the specification is. In the case of env shells the last one takes priority because of how shell processes them.

aryanjassal commented:
After a discussion with Brian, we have come to the following conclusion:

Instead of using a separate option like `--preserve-newline`, then merging the secrets passed in to that and the original arguments, we instead pass in all the secret paths as arguments. The flag can be used as a modifier to the argument list instead of acting as a separate argument list to keep track of.

This means that all secrets specified are, by default, trimmed. However, if a secret name is present in the list to preserve newlines, then the newlines will be preserved. This approach does cause duplication to the secret path, but that's not a huge concern, as most of the times, the newlines would be trimmed as usual.

This approach solves all the issues. The newline preservation takes precedence. If the newlines must be trimmed, then don't specify it to be preserved. As the arguments need to be populated anyways, they will never be empty, and the usual checks can be done from within the parser without need of manually creating an empty object or throwing an error from the action.

This also, to an extent, treats the symptoms of `'--'` being weird to use. That is an exceptionally weird issue, as they work perfectly in the tests, but break when used via the command line. I can start by using `shell` in `pkExec`, which is being used in the tests. This will let me find out if the shell is misbehaving or not.

CMCDragonkai commented:
Can you also review https://github.com/MatrixAI/Polykey/issues/222#issuecomment-2452549365? I want to make sure we're laying foundations to enable egress schema implementation.

aryanjassal commented:
> Can you also review [MatrixAI/Polykey#222 (comment)](https://github.com/MatrixAI/Polykey/issues/222#issuecomment-2452549365)? I want to make sure we're laying foundations to enable egress schema implementation.

Just to confirm, this egress schema refers to a schema which would allow formatting of the secrets in a particular way at the time of egress, which would be using the `secrets env` command in this case? For example, the schema would specify if a secret needs to preserve trailing newline or not. If so, then yes, foundation for supporting this is being laid in this PR.

aryanjassal commented on src/utils/parsers.ts:
```diff
@@ -208,6 +208,8 @@ function parseEnvArgs(
+  console.error('cur', current)
   if (current[1].length === 0) {
     // Parse a secret path
     if (value !== '--') {
```
For now, we are also returning the `'--'` in the returned command arguments as such, but it doesn't seem ideal.

```ts
> // Command setup
> // Calling with `polykey secrets env vault:SECRET -- command`
> console.log(args);
[['vault', 'SECRET', undefined'], ['--', 'command']]
```

Currently, this is being removed at the beginning of the `env` command's action, but this shouldn't be the case in the first place. How should we deal with it, if at all?

One idea I had was to store some state inside `prev` as a JSON object, so we can extract the args. The parser can set and get the previously-set state to switch contexts between parsing vaults to parsing commands.

Another one is to store this state out of the function, so the `prev` doesn't get messed up, but storing the state of a function outside the actual function just feels weird.

Should this be addressed, or the command removing `'--'` is fine and the behaviour doesn't need modification?

aryanjassal replied:
We can't rely on any checks like `if (current[1].length === 1) { current[1].shift(); }`, as once it has been removed, it will trigger the condition again, and if the user intends to pass `'--'` as an actual value, then this approach will also incorrectly remove that, too. As the array is of strings, I cannot use values like `null` or `undefined` as placeholder values to be removed later, and I also can't use an empty string as the placeholder because `commander` can return any combination of a string.

aryanjassal replied:
To address this, we are storing a state flag in the parser `prev`. If we encounter the `--`, then we set the flag and continue. This will then result in the following iterations appending to the command list instead. We can simply ignore this state when deconstructing the array anywhere else.

aryanjassal commented on tests/secrets/env.test.ts:
```diff
@@ -694,7 +721,40 @@ describe('commandEnv', () => {
+    await polykeyAgent.vaultManager.withVaults([vaultId], async (vault) => {
+      await vaultOps.addSecret(vault, secretName, secretContent);
+    });
+    const command = [
+      'secrets',
+      'env',
+      '-np',
+      dataDir,
+      '--env-format',
+      'unix',
+      '--preserve-newline',
+      `${vaultName}:${secretName}`,
+      `${vaultName}:${secretName}`,
+      '--',
+      'node',
+      '-e',
+      'console.log(JSON.stringify(process.env))',
+    ];
```
Very interestingly, if the order of the flag is changed to the following, the flag gets parsed as a vault name instead. I tried to replicate this behaviour on the command line manually, but that didn't work, and it worked as expected.

```ts
const command = [
  ...
  `${vaultName}:${secretName}`,
  '--preserve-newline',
  `${vaultName}:${secretName}`,
  ...
];
```

First, for the tests, specifying a single `'--'` works but shell needs two of them, now this inconsistency. Something weird is going on. I will investigate this when I work next.

aryanjassal replied:
So, apparently, this issue is resolved by removing the `this.passThroughOptions()`, which makes the tests work as on the CLI, but that has the side-effect of the tests also needing the double `'--'` to be parsed properly. I think we might need to rely on `process.argv` to process the secret paths and the commands separately.

tegefaulkes replied:
Were the `--preserve-newline` options being set before the arguments of the command?  The comment above seems to suggest that you were setting the options after the arguments. Which `passThroughOptions` will treat as arguments because it's telling commander to treat everything after the first argument as more aguments.

aryanjassal replied:
Interestingly enough, doing it like that actually works on the CLI, and the CLI needs two `'--'`, as if `passThroughOptions()` isn't having an effect on the actual CLI of the command, but just the tests.

If `passThroughOptions()` doesn't even work on the CLI, and only on the tests (as what we can see here), should I just remove it and look for alternative solutions for the `'--'` issue, like trying to use `process.argv` for arguments and parsing instead of `commander`?

I enabled the shell mode in `pkExec` in the tests, and the behaviour was consistent with the actual shell, in that I needed two of `'--'` and the options were working in any order.

aryanjassal replied:
After digging into the reason for needing double `--` and the inconsistency between the tests and the CLI, I discovered that I needed an option, `this.enablePositionalOptions()` to be set. Without it, I cannot pass through the options and they are parsed as normal.

However, enabling the option breaks other things, especially the option availability on all subcommand layers. That doesn't work, and my tries of using ChatGPT to fix it have succeeded a little, but ultimately failed to fix all the issues.

I will make an issue on `commanderjs` repo, asking them to add a feature to allow this, or provide a workaround.

aryanjassal replied:
Until then, to run commands, just use double `--` instead of a single one.

```console
[aryanj@matrix-34xx:~]$ polykey secrets env vault:SECRET -- -- bash -c 'echo $SECRET'
this is the secret contents
```

aryanjassal commented:
Key changes in this PR:
- Added a new error which is thrown whenever a subprocess fails. It is used in `secrets env` (if the command fails) and `secrets edit` (if the editor fails).
- Properly removed handlers from editor process in `secrets edit`
- Automatically trimming newlines when exporting secrets using `secrets env`
- Preserving newlines of any secrets listed in `-pn` option. (should there be a short form of this option?)
- Created type for parsed secret path and value (`[string, string?, string?]` like `[vaultName, secretPath, value]`)
- Added tests for trimming and preserving newlines using fastcheck

Need input on:
- Is the addition and usage of `ErrorPolykeyCLISubprocessFailure` reasonable?
- Should `-pn --preserve-newline` have a short form of the option?

@tegefaulkes

CMCDragonkai commented:
> Key changes in this PR:
> 
> * Added a new error which is thrown whenever a subprocess fails. It is used in `secrets env` (if the command fails) and `secrets edit` (if the editor fails).
> * Properly removed handlers from editor process in `secrets edit`
> * Automatically trimming newlines when exporting secrets using `secrets env`
> * Preserving newlines of any secrets listed in `-pn` option. (should there be a short form of this option?)
> * Created type for parsed secret path and value (`[string, string?, string?]` like `[vaultName, secretPath, value]`)
> * Added tests for trimming and preserving newlines using fastcheck
> 
> Need input on:
> 
> * Is the addition and usage of `ErrorPolykeyCLISubprocessFailure` reasonable?
> * Should `-pn --preserve-newline` have a short form of the option?
> 
> @tegefaulkes

If we are using exec, subprocess failure should be temporary on Unix systems, because of process replacement. It would be more accurate to say that you were not able to execute the subprocess. Also I prefer the name ChildProcess.

On windows however there is no process replacement. It's always a child process. Did you do cross platform sanity checks?

aryanjassal commented:
> Did you do cross platform sanity checks?

Oh, it completely slipped my mind. I will do this today before merging.

tegefaulkes commented on src/secrets/CommandEnv.ts:
```diff
@@ -160,7 +157,27 @@ class CommandEnv extends CommandPolykey {
+
+              // Find if we need to preserve the newline for this secret
+              let preserveNewline = false;
+              if (options.preserveNewline) {
+                for (const pair of options.preserveNewline) {
+                  if (
+                    pair[0] === nameOrId &&
+                    (pair[1] === newName || pair[1] == null)
+                  ) {
+                    preserveNewline = true;
+                    break;
+                  }
+                }
+              }
```
This is poorly done. 

First we only ever do `if(thing)` if thing is a boolean. Otherwise you need to be stricter about it. In this case we're checking if `options.preserveNewLine` is defined. We always check this with `if(options.preserveNewLine != null)`. 

Secondly, if you're checking if something is within a set, you need to use a map rather than iterate over the whole set. I think you're checking if a path or a whole vault for that path is preserving the newline here? That should have a comment explaining that as well. But for the most part rather than iterate over the whole set of `options.preserveNewLine` You should structure it as a map and use that instead. To check if something is in a map is log complexity vs the linear complexity of iterating over a list.

aryanjassal replied:
> Secondly, if you're checking if something is within a set, you need to use a map rather than iterate over the whole set. I think you're checking if a path or a whole vault for that path is preserving the newline here? That should have a comment explaining that as well. But for the most part rather than iterate over the whole set of `options.preserveNewLine` You should structure it as a map and use that instead. To check if something is in a map is log complexity vs the linear complexity of iterating over a list.

I'm not sure how to achieve this. `options.preserveNewline` is an array of parsed environment paths like the ones parsed in the actual arguments. (like `[ ['vaultName', 'secretPath', 'value'] ]`). Perhaps the naming being 'pair' is kind of misleading. How would I structure this in a map?

Do you mean doing something like putting vault name as the key, and all the provided paths as the values, then use that for the checks here?

tegefaulkes replied:
```ts
const set = new Set<string>();

// The options.preserveNewline fills the set.
for (const [vaultName, path] of options.preserveNewLine) {
  set.add(`${vaultName}${path != null ? `:${path}` : ``}`)
}

// Then when checking you can check the path or the vault name.
function checkPreserveNewline(vaultName: string, path: string): boolean {
  const preserveVault = set.has(vaultName);
  const preservePath = set.has(`${vaultName}:${path}`);
  return preserveVault || preservePath;
}

for (const [vaultName, path] of secrets) {
  if (checkPreserveNewline(vaultName, path)) {
    // Do the thing.
  }
}
```

Can you guess why this would be preferred?

tegefaulkes commented on src/secrets/CommandEnv.ts:
```diff
@@ -175,14 +192,26 @@ class CommandEnv extends CommandPolykey {
+            // If a cmd is provided then we default to exec it
             switch (platform) {
               case 'linux':
-              // Fallthrough
```
Why did you remove the `// Fallthrough` comment? This is like a `this page is intentionally left blank`. With switch case statements a common problem is forgetting the `break:` between cases, this is letting people know it's intentional and not a mistake.

aryanjassal replied:
Oh. I just did it for clarity, as in other places, the fallthrough comment has been omitted. Especially in error composition in Polykey, Polykey-CLI, and js-rpc. I was just being consistent with that. I can change this back if needed, but how do I handle the other switch statements with that?

tegefaulkes replied:
Add it back in. If you see it missing in other places then add it in there. I thought the linter complains about this but it might just be my IDE.

aryanjassal replied:
Yeah, I think your IDE does some extra processing over the regular standards we have. Can you take a look at `Polykey:src/network/utils.ts:487` and see if your IDE still complains for that? If so, then there are a couple places where we would need to add the fallthrough comment.

This has also been done in `Polykey-CLI:src/utils/utils.ts:194`, and adding comments would make it more difficult to read.

```ts
switch (lastChar) {
  case 'n':
    return '\n';
  case 'r':
    return '\r';
  case 't':
    return '\t';
  case 'v':
    return '\v';
  case 'f':
    return '\f';
  case '"':
  case "'":
  case '`':
  case '\\':
    return lastChar;
}
// versus
switch (lastChar) {
  case 'n':
    return '\n';
  case 'r':
    return '\r';
  case 't':
    return '\t';
  case 'v':
    return '\v';
  case 'f':
    return '\f';
  case '"':
  // Fallthrough
  case "'":
  // Fallthrough
  case '`':
  // Fallthrough
  case '\\':
    return lastChar;
}

```

tegefaulkes replied:
Seems my IDE doesn't complain about that. All the same messing up a fallthrough in a switch case is a common bug. and annoying to debug because intention can be unclear. Its best to be explicit about the fallthough.

If you're worried about readability then just put the comment on the same line
```ts
switch (thing) {
// ...
    case '"': // Fallthrough
    case "'": // Fallthrough
    case '`': // Fallthrough
// ...
}
```

aryanjassal replied:
Okay, yeah, this looks clean and is also informative. Ill make this change across all relevant cases in this repo.

tegefaulkes commented on tests/utils/utils.ts:
```diff
@@ -112,6 +112,13 @@ const cmdArgsArrayArb = fc
   })
   .noShrink();
 
+const vaultNameArb = fc
+  .stringOf(
+    fc.char().filter((c) => binParsers.vaultNameRegex.test(c)),
+    { minLength: 1, maxLength: 100 },
+  )
+  .noShrink();
```
You should generate the string directly from the regex here.

aryanjassal replied:
If I try to do this with `stringMatching`, then the tests fail to run with the following error:

```
Assertions of kind Lookahead not implemented yet!

      113 |   .noShrink();
      114 |
    > 115 | const vaultNameArb = fc.stringMatching(binParsers.vaultNameRegex).noShrink();
```

As such, I will retain the previous approach of checking individual characters and concatenating them.

aryanjassal replied:
The previous approach looked very bad and also worked kinda weirdly. So, I have updated the code to utilise this instead of the previous approach.

```ts
const vaultNameArb = fc
  .string({ minLength: 1, maxLength: 100 })
  .filter((str) => binParsers.vaultNameRegex.test(str))
  .noShrink();
```

aryanjassal commented:
> On windows however there is no process replacement. It's always a child process. Did you do cross platform sanity checks?

https://github.com/MatrixAI/Polykey-CLI/issues/331

aryanjassal commented:
The CI is passing, and the PR has been approved. I will be merging it now.

CMCDragonkai commented on src/errors.ts:
```diff
@@ -84,6 +84,11 @@ class ErrorPolykeyCLIUnexpectedError<T> extends ErrorPolykeyCLI<T> {
   exitCode = sysexits.SOFTWARE;
 }
 
+class ErrorPolykeyCLISubprocessFailure<T> extends ErrorPolykeyCLI<T> {
+  static description = 'A subprocess failed to exit gracefully';
+  exitCode = sysexits.UNKNOWN;
+}
+
```
Can you change the description and name according to https://github.com/MatrixAI/Polykey-CLI/pull/329#issuecomment-2481400933

aryanjassal replied:
I had actually merged this PR and had forgotten to make this change. So, I made this change in another PR I was working on. This error has been renamed to `ErrorPolykeyCLIChildProcessFailure` with the updated description in https://github.com/MatrixAI/Polykey-CLI/pull/336.

CMCDragonkai replied:
Is this not specific to the env command? If it's called `ErrorPolykeyCLIChildProcessFailure` that would imply in general the idea of a child process failure to PK CLI.

Unless you're also using this in any situation you're going to create a child process?

aryanjassal replied:
This error is not specific to the `env` command. As `secrets edit` also spawns the editor in a child process, any failure of the child process is wrapped in this error.

So, yes, in any situation where a child process might fail, this error should be used to wrap it, unless there is a better error in that context.

CMCDragonkai commented on src/identities/CommandDiscover.ts:
```diff
@@ -103,8 +103,8 @@ class CommandDiscover extends CommandPolykey {
                     break;
-                  case 'processed':
-                  case 'cancelled':
+                  case 'processed': // Fallthrough
+                  case 'cancelled': // Fallthrough
```
Why need these comments?

aryanjassal replied:
They are not needed, but are recommended to be inserted. This will ensure that the case is intentionally left blank as a fallthrough and is not a mistake, where it was intended to fill in the content later.

aryanjassal replied:
The basic linter (eslint) doesn't care, but IDEs (like WebStorm) actually report this as an error if there is a blank case without a fallthrough comment.

CMCDragonkai commented:
This was a significant change to the UX of secrets env. Now it is actually usable.
